File: codetraverse/ast_diff/TSdiff.py
import json
from tree_sitter import Language, Parser, Node
import tree_sitter_typescript
from .Detailedchanges import DetailedChanges
from .basefilediff import BaseFileDiff

class TypeScriptFileDiff(BaseFileDiff):
    """Analyzes and compares two TypeScript ASTs for semantic differences."""
    def __init__(self, module_name=""):
        self.changes = DetailedChanges(module_name)

    def get_decl_name(self, node: Node) -> str:
        """Finds the name of a TypeScript declaration."""
        # Handle `export` statements by looking inside them
        if node.type == 'export_statement':
            if node.named_child_count > 0:
                declaration_node = node.named_child(node.named_child_count - 1)
                if declaration_node:
                    return self.get_decl_name(declaration_node)

        # Handle `const myFunc = () => {}` or `const myVar = 1;`
        if node.type == 'lexical_declaration':
            # A lexical declaration can have multiple declarators, like `const a = 1, b = 2;`
            # We'll just grab the name from the first one for simplicity.
            declarator_node = node.named_child(0)
            if declarator_node and declarator_node.type == 'variable_declarator':
                name_node = declarator_node.child_by_field_name('name')
                if name_node:
                    return name_node.text.decode('utf8')

        # Handle `function`, `class`, `interface`, etc.
        name_node = node.child_by_field_name('name')
        if name_node:
            return name_node.text.decode('utf8')

        return None

    def extract_components(self, root: Node):
        """Extracts all top-level declarations from a TypeScript AST, including functions inside objects."""
        functions, classes, interfaces, types, enums, constants = {}, {}, {}, {}, {}, {}
        
        node_type_map = {
            'function_declaration': functions,
            'class_declaration': classes,
            'interface_declaration': interfaces,
            'type_alias_declaration': types,
            'enum_declaration': enums,
        }
        
        declarations = root.children if root.type == 'program' else []

        for child in declarations:
            node_to_process = child
            
            # Look inside `export` statements
            if child.type == 'export_statement':
                if child.named_child_count > 0:
                    declaration_node = child.named_child(child.named_child_count - 1)
                    if declaration_node:
                        node_to_process = declaration_node
            
            node_type = node_to_process.type

            if node_type == 'lexical_declaration':
                declarator = node_to_process.named_child(0)
                if declarator and declarator.type == 'variable_declarator':
                    name = self.get_decl_name(child)
                    value_node = declarator.child_by_field_name('value')
                    
                    if name and value_node:
                        # Handle `as const` expressions by unwrapping them
                        actual_value_node = value_node
                        if actual_value_node.type == 'as_expression' and actual_value_node.child_count > 0:
                            actual_value_node = actual_value_node.child(0)

                        # Case 1: const is an arrow function
                        if actual_value_node.type == 'arrow_function':
                            data_tuple = (child, child.text.decode(errors="ignore"), child.start_point, child.end_point)
                            functions[name] = data_tuple
                        
                        # Case 2: const is an object, which might contain functions (common in .tsx)
                        elif actual_value_node.type == 'object':
                            # The object itself is a constant
                            data_tuple = (child, child.text.decode(errors="ignore"), child.start_point, child.end_point)
                            constants[name] = data_tuple
                            
                            # Now, look for functions inside the object's properties
                            for pair_node in actual_value_node.children:
                                if pair_node.type == 'pair':
                                    key_node = pair_node.child_by_field_name('key')
                                    val_node = pair_node.child_by_field_name('value')
                                    
                                    if key_node and val_node and val_node.type == 'arrow_function':
                                        inner_func_name = f"{name}.{key_node.text.decode('utf8')}"
                                        inner_data_tuple = (pair_node, pair_node.text.decode(errors="ignore"), pair_node.start_point, pair_node.end_point)
                                        functions[inner_func_name] = inner_data_tuple
                        
                        # Case 3: const is something else
                        else:
                            data_tuple = (child, child.text.decode(errors="ignore"), child.start_point, child.end_point)
                            constants[name] = data_tuple

            elif node_type in node_type_map:
                name = self.get_decl_name(child)
                if name:
                    target_dict = node_type_map[node_type]
                    target_dict[name] = (child, child.text.decode(errors="ignore"), child.start_point, child.end_point)

        return functions, classes, interfaces, types, enums, constants

    def diff_components(self, before_map: dict, after_map: dict):
        """Compares two dictionaries of components and returns the diff."""
        before_names = set(before_map.keys())
        after_names = set(after_map.keys())
        added_names = after_names - before_names
        deleted_names = before_names - after_names
        common_names = before_names & after_names
        
        added = [(n, after_map[n][1], {"start": after_map[n][2], "end": after_map[n][3]}) for n in sorted(added_names)]
        deleted = [(n, before_map[n][1], {"start": before_map[n][2], "end": before_map[n][3]}) for n in sorted(deleted_names)]
        modified = []
        for name in sorted(common_names):
            _, old_body, old_start, old_end = before_map[name]
            _, new_body, new_start, new_end = after_map[name]
            if old_body.strip() != new_body.strip():
                 modified.append((name, old_body, new_body, {"old_start": old_start, "old_end": old_end, "new_start": new_start, "new_end": new_end}))
        return {"added": added, "deleted": deleted, "modified": modified}

    def compare_two_files(self, old_file_ast: Node, new_file_ast: Node) -> DetailedChanges:
        """The main method to compare two TypeScript files."""
        old_funcs, old_classes, old_ifaces, old_types, old_enums, old_consts = self.extract_components(old_file_ast.root_node)
        new_funcs, new_classes, new_ifaces, new_types, new_enums, new_consts = self.extract_components(new_file_ast.root_node)

        # Define what we're comparing
        category_map = {
            "functions": (old_funcs, new_funcs),
            "classes": (old_classes, new_classes),
            "interfaces": (old_ifaces, new_ifaces),
            "types": (old_types, new_types),
            "enums": (old_enums, new_enums),
            "constants": (old_consts, new_consts),
        }

        # Run diff and record changes
        for category, (old_map, new_map) in category_map.items():
            diff = self.diff_components(old_map, new_map)
            for change_type in ["added", "deleted", "modified"]:
                for item in diff[change_type]:
                    self.changes.add_change(category, change_type, item)

        return self.changes
        
    def process_single_file(self, file_ast: Node, mode="deleted") -> DetailedChanges:
        """Processes a single file that was either entirely added or deleted."""
        funcs, classes, interfaces, types, enums, consts = self.extract_components(file_ast.root_node)

        category_map = {
            "functions": funcs,
            "classes": classes,
            "interfaces": interfaces,
            "types": types,
            "enums": enums,
            "constants": consts,
        }

        for category, component_map in category_map.items():
            for name, data_tuple in component_map.items():
                # data_tuple is (node, text, start_point, end_point)
                item = (name, data_tuple[1], {"start": data_tuple[2], "end": data_tuple[3]})
                self.changes.add_change(category, mode, item)
        
        return self.changes

File: codetraverse/ast_diff/TSdiff.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypeScriptFileDiff = void 0;
var Detailedchanges_1 = require("./Detailedchanges");
var basefilediff_1 = require("./basefilediff");
var TypeScriptFileDiff = /** @class */ (function (_super) {
    __extends(TypeScriptFileDiff, _super);
    function TypeScriptFileDiff(moduleName) {
        if (moduleName === void 0) { moduleName = ""; }
        var _this = _super.call(this, moduleName) || this;
        _this.changes = new Detailedchanges_1.DetailedChanges(moduleName);
        return _this;
    }
    TypeScriptFileDiff.prototype.getDeclName = function (node) {
        if (node.type === 'export_statement') {
            if (node.namedChildCount > 0) {
                var declarationNode = node.namedChildren[node.namedChildren.length - 1];
                if (declarationNode) {
                    return this.getDeclName(declarationNode);
                }
            }
        }
        if (node.type === 'lexical_declaration') {
            var declaratorNode = node.namedChildren[0];
            if (declaratorNode && declaratorNode.type === 'variable_declarator') {
                var nameNode_1 = declaratorNode.childForFieldName('name');
                if (nameNode_1) {
                    return nameNode_1.text;
                }
            }
        }
        var nameNode = node.childForFieldName('name');
        if (nameNode) {
            return nameNode.text;
        }
        return null;
    };
    TypeScriptFileDiff.prototype.extractClassMethods = function (classNode, className, functionsDict, fieldsDict) {
        var classBody = classNode.children.find(function (c) { return c.type === 'class_body'; });
        if (!classBody) {
            return;
        }
        for (var _i = 0, _a = classBody.children; _i < _a.length; _i++) {
            var child = _a[_i];
            if (child.type === 'method_definition') {
                var methodNameNode = child.childForFieldName('name');
                if (methodNameNode) {
                    var methodName = methodNameNode.text;
                    var qualifiedName = "".concat(className, ".").concat(methodName);
                    functionsDict[qualifiedName] = [child, child.text, child.startPosition, child.endPosition];
                }
            }
            else if (child.type === 'public_field_definition') {
                var fieldNameNode = child.childForFieldName('name');
                if (fieldNameNode) {
                    var fieldName = fieldNameNode.text;
                    var qualifiedName = "".concat(className, ".").concat(fieldName);
                    fieldsDict[qualifiedName] = [child, child.text, child.startPosition, child.endPosition];
                }
            }
        }
    };
    TypeScriptFileDiff.prototype.extractComponents = function (root) {
        var functions = {};
        var classes = {};
        var interfaces = {};
        var types = {};
        var enums = {};
        var constants = {};
        var fields = {};
        var nodeTypeMap = {
            'function_declaration': functions,
            'class_declaration': classes,
            'interface_declaration': interfaces,
            'type_alias_declaration': types,
            'enum_declaration': enums,
        };
        var declarations = root.type === 'program' ? root.children : [];
        for (var _i = 0, declarations_1 = declarations; _i < declarations_1.length; _i++) {
            var child = declarations_1[_i];
            var nodeToProcess = child;
            if (child.type === 'export_statement') {
                if (child.namedChildCount > 0) {
                    var declarationNode = child.namedChildren[child.namedChildren.length - 1];
                    if (declarationNode) {
                        nodeToProcess = declarationNode;
                    }
                }
            }
            var nodeType = nodeToProcess.type;
            if (nodeType === 'lexical_declaration') {
                var declarator = nodeToProcess.namedChildren[0];
                if (declarator && declarator.type === 'variable_declarator') {
                    var name_1 = this.getDeclName(child);
                    var valueNode = declarator.childForFieldName('value');
                    if (name_1 && valueNode) {
                        var actualValueNode = valueNode;
                        if (actualValueNode.type === 'as_expression' && actualValueNode.childCount > 0) {
                            actualValueNode = actualValueNode.children[0];
                        }
                        if (actualValueNode.type === 'arrow_function') {
                            functions[name_1] = [child, child.text, child.startPosition, child.endPosition];
                        }
                        else if (actualValueNode.type === 'object') {
                            constants[name_1] = [child, child.text, child.startPosition, child.endPosition];
                            for (var _a = 0, _b = actualValueNode.children; _a < _b.length; _a++) {
                                var pairNode = _b[_a];
                                if (pairNode.type === 'pair') {
                                    var keyNode = pairNode.childForFieldName('key');
                                    var valNode = pairNode.childForFieldName('value');
                                    if (keyNode && valNode && valNode.type === 'arrow_function') {
                                        var innerFuncName = "".concat(name_1, ".").concat(keyNode.text);
                                        functions[innerFuncName] = [pairNode, pairNode.text, pairNode.startPosition, pairNode.endPosition];
                                    }
                                }
                            }
                        }
                        else {
                            constants[name_1] = [child, child.text, child.startPosition, child.endPosition];
                        }
                    }
                }
            }
            else if (nodeType in nodeTypeMap) {
                var name_2 = this.getDeclName(child);
                if (name_2) {
                    var targetDict = nodeTypeMap[nodeType];
                    targetDict[name_2] = [child, child.text, child.startPosition, child.endPosition];
                    if (nodeType === 'class_declaration') {
                        this.extractClassMethods(nodeToProcess, name_2, functions, fields);
                    }
                }
            }
        }
        return [functions, classes, interfaces, types, enums, constants, fields];
    };
    TypeScriptFileDiff.prototype.diffComponents = function (beforeMap, afterMap) {
        var beforeNames = new Set(Object.keys(beforeMap));
        var afterNames = new Set(Object.keys(afterMap));
        var addedNames = Array.from(afterNames).filter(function (name) { return !beforeNames.has(name); });
        var deletedNames = Array.from(beforeNames).filter(function (name) { return !afterNames.has(name); });
        var commonNames = Array.from(beforeNames).filter(function (name) { return afterNames.has(name); });
        var added = addedNames.sort().map(function (n) { return [n, afterMap[n][1], { start: afterMap[n][2], end: afterMap[n][3] }]; });
        var deleted = deletedNames.sort().map(function (n) { return [n, beforeMap[n][1], { start: beforeMap[n][2], end: beforeMap[n][3] }]; });
        var modified = [];
        for (var _i = 0, _a = commonNames.sort(); _i < _a.length; _i++) {
            var name_3 = _a[_i];
            var _b = beforeMap[name_3], oldAst = _b[0], oldBody = _b[1], oldStart = _b[2], oldEnd = _b[3];
            var _c = afterMap[name_3], newAst = _c[0], newBody = _c[1], newStart = _c[2], newEnd = _c[3];
            if (oldBody.trim() !== newBody.trim()) {
                modified.push([name_3, oldBody, newBody, { old_start: oldStart, old_end: oldEnd, new_start: newStart, new_end: newEnd }]);
            }
        }
        return { added: added, deleted: deleted, modified: modified };
    };
    TypeScriptFileDiff.prototype.compareTwoFiles = function (oldFileAst, newFileAst) {
        var _a = this.extractComponents(oldFileAst), oldFuncs = _a[0], oldClasses = _a[1], oldIfaces = _a[2], oldTypes = _a[3], oldEnums = _a[4], oldConsts = _a[5], oldFields = _a[6];
        var _b = this.extractComponents(newFileAst), newFuncs = _b[0], newClasses = _b[1], newIfaces = _b[2], newTypes = _b[3], newEnums = _b[4], newConsts = _b[5], newFields = _b[6];
        var categoryMap = {
            "functions": [oldFuncs, newFuncs],
            "classes": [oldClasses, newClasses],
            "interfaces": [oldIfaces, newIfaces],
            "types": [oldTypes, newTypes],
            "enums": [oldEnums, newEnums],
            "constants": [oldConsts, newConsts],
            "fields": [oldFields, newFields],
        };
        for (var category in categoryMap) {
            var _c = categoryMap[category], oldMap = _c[0], newMap = _c[1];
            var diff = this.diffComponents(oldMap, newMap);
            for (var changeType in diff) {
                for (var _i = 0, _d = diff[changeType]; _i < _d.length; _i++) {
                    var item = _d[_i];
                    this.changes.add_change(category, changeType, item);
                }
            }
        }
        return this.changes;
    };
    TypeScriptFileDiff.prototype.processSingleFile = function (fileAst, mode) {
        if (mode === void 0) { mode = "deleted"; }
        var _a = this.extractComponents(fileAst), funcs = _a[0], classes = _a[1], interfaces = _a[2], types = _a[3], enums = _a[4], consts = _a[5], fields = _a[6];
        var categoryMap = {
            "functions": funcs,
            "classes": classes,
            "interfaces": interfaces,
            "types": types,
            "enums": enums,
            "constants": consts,
            "fields": fields,
        };
        for (var category in categoryMap) {
            var componentMap = categoryMap[category];
            for (var name_4 in componentMap) {
                var dataTuple = componentMap[name_4];
                var item = [name_4, dataTuple[1], { start: dataTuple[2], end: dataTuple[3] }];
                this.changes.add_change(category, mode, item);
            }
        }
        return this.changes;
    };
    return TypeScriptFileDiff;
}(basefilediff_1.BaseFileDiff));
exports.TypeScriptFileDiff = TypeScriptFileDiff;
